cirq/work/collector.py

```diff
@@ -0,0 +1,164 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+from typing import Optional, Any, Iterable, Union, List
+
+import numpy as np
+
+from cirq import circuits, study, value
+from cirq.work import sampler, work_pool
+
+
+@value.value_equality(unhashable=True)
+class CircuitSampleJob:
+    """Describes a sampling task."""
+
+    # pylint: disable=redefined-builtin
+    def __init__(self,
+                 circuit: circuits.Circuit,
+                 *,
+                 repetitions: int,
+                 id: Any = None):
+        """
+        Args:
+            circuit: The circuit to sample from.
+            repetitions: How many times to sample the circuit.
+            id: An arbitrary value associated with the job. This value is used
+                so that when a job completes and is handed back, it is possible
+                to tell what the job was for. For example, the key could be a
+                string like "main_run" or "calibration_run", or it could be set
+                to the component of the Hamiltonian (e.g. a PauliString) that
+                the circuit is supposed to be helping to estimate.
+        """
+        self.circuit = circuit
+        self.repetitions = repetitions
+        self.id = id
+    # pylint: enable=redefined-builtin
+
+    def _value_equality_values_(self):
+        return self.circuit, self.repetitions, self.id
+
+    def __repr__(self):
+        return ('cirq.CircuitSampleJob('
+                'id={!r}, repetitions={!r}, circuit={!r})').format(
+                    self.id, self.repetitions, self.circuit)
+
+
+class SampleCollector(metaclass=abc.ABCMeta):
+    """An interface for concurrently collecting sample data."""
+
+    @abc.abstractmethod
+    def next_job(self
+                ) -> Union[None, CircuitSampleJob, Iterable[CircuitSampleJob]]:
+        """Called by driving code when more sampling can be started.
+
+        Returns:
+            A CircuitSampleJob describing the circuit to sample, how many
+            samples to take, and a key value that can be used in the
+            `on_job_result` method to recognize which job this is.
+        """
+
+    @abc.abstractmethod
+    def on_job_result(self, job: CircuitSampleJob,
+                      result: study.TrialResult) -> None:
+        """Called by driving code when sample results have become available.
+
+        The results should be incorporated into the collector's state.
+        """
+
+
+async def async_collect_samples(collector: SampleCollector,
```

Comment:
Consider making this an instance method on `SampleCollector`?

Reply:
Done.